File: Leetcode/C++/2187. Minimum Time to Complete Trips/solution.cpp
// https://leetcode.com/problems/minimum-time-to-complete-trips/description/

#include <vector>
#include <algorithm>

using namespace std;

class Solution
{
public:
    long long minimumTime(vector<int> &time, int totalTrips)
    {
        sort(time.begin(), time.end());
        long long int start = 0, end = (long long int)time[0] * totalTrips;
        while (start < end)
        {
            long long int mid = start + (end - start) / 2; // Prevent overflow
            long long int currentTrips = 0;
            for (auto bus : time)
            {
                currentTrips += mid / bus;
            }
            // If currentTrips smallest than totalTrips mid is cant be answer so move start to mid + 1
            if (currentTrips < totalTrips)
            {
                start = mid + 1;
            }
            // If currentTrips greater than or equal totalTrips mid can answer so move end to mid
            else
            {
                end = mid;
            }
        }
        return start; // also end is answer too
    }
};

File: Leetcode/Python3/2187. Minimum Time to Complete Trips/solution.py
# https://leetcode.com/problems/minimum-time-to-complete-trips/


class Solution:
    def minimumTime(self, time: List[int], totalTrips: int) -> int:
        time.sort()
        # Max time == smallest time * totalTrips
        start, end = 0, time[0] * totalTrips
        while start < end:
            mid = start + (end - start) // 2    # mid == time
            cur_trips = 0
            for bus in time:
                cur_trips += mid // bus
            if cur_trips < totalTrips:  # If current trips small than totalTrips it cant be answer so move start to mid + 1
                start = mid + 1
            else:   # Current trips is greater than or equal totalTrips mid is can be answer so move end to mid
                end = mid
        return end  # start is also the answer
